server/controllers/finance/accounts/extra.js

```diff
@@ -0,0 +1,127 @@
+ * @returns Promise - promise wrapping an integer periodId
+ */
+function getPeriodForDate(date) {
+  const sql = `
+    SELECT id FROM period WHERE start_date <= DATE(?) AND end_date >= DATE(?);
+  `;
+
+  return db.one(sql, [date, date])
+    .then(data => data.id);
+}
+
+/**
+ * @function getPeriodAccountBalanceUntilDate
+ * @private
+ *
+ * @description
+ * Sums the balance of an account up to just before the given date's period.  For example, if the date was provided as
+ * May 19, 2016, this function would return the account balance at the end of April 30, 2016.  This is useful for
+ * computing opening balances of the period given a date.
+ *
+ * @param {Number} accountId - the account_id for the period_total table
+ * @param {Date} date - the upper limit of the period
+ * @param {Number} fiscalYearId - the fiscal_year_id for the period_total table
+ *
+ * @returns Promise - promise wrapping the balance object
+ */
+function getPeriodAccountBalanceUntilDate(accountId, date, fiscalYearId) {
+  const sql = `
+    SELECT SUM(debit - credit) AS balance
+    FROM period_total JOIN period ON period.id = period_total.period_id
+    WHERE period_total.account_id = ?
+      AND period.end_date <= DATE(?)
+      AND period.fiscal_year_id = ?;
+  `;
+
+  return db.one(sql, [accountId, date, fiscalYearId])
+    .then(data => data.balance);
+}
+
+/**
+ * @function getComputedAccountBalanceUntilDate
+ * @private
+ *
+ * @description
+ * Sums general ledger lines hitting an account during a period, up to (and including) the provided date.
+ */
+function getComputedAccountBalanceUntilDate(accountId, date, periodId) {
+  const sql = `
+    SELECT SUM(debit_equiv - credit_equiv) AS balance FROM general_ledger
+    WHERE account_id = ?
+      AND trans_date <= DATE(?)
+      AND period_id = ?;
+  `;
+
+  return db.one(sql, [accountId, date, periodId])
+    .then(data => data.balance);
+}
+
+
+/**
+ * @method getOpeningBalanceForDate
+ * @public
+ *
+ * @description
+ * Query the database for an account's balance as of a start date.  Note that the date should be escaped (via new
+ * Date()) prior to calling this function - this method is expected to does not do any escaping.
+ *
+ * @param {Number} accountId - the identifier for the account
+ * @param {Date} date - the date that the opening balance should be computed through
+ * @returns {Promise} - promise wrapping the balance of the account
+ */
+function getOpeningBalanceForDate(accountId, date) {
+  let balance = 0;
+
+  return getFiscalYearForDate(date)
+    .then(fiscalYearId =>
+      getPeriodAccountBalanceUntilDate(accountId, date, fiscalYearId)
+    )
+    .then((previousPeriodClosingBalance) => {
+      balance += previousPeriodClosingBalance;
+      return getPeriodForDate(date);
+    })
+    .then(periodId =>
+      getComputedAccountBalanceUntilDate(accountId, date, periodId)
+    )
+    .then(runningPeriodBalance => balance + runningPeriodBalance);
```

Comment:
In theory, the currency filter takes care of that.  However, I could make them into fixed numbers if you think it's best.